api_tests/resources/views/test_resource_detail.py

```diff
@@ -131,28 +166,346 @@ def test_anonymized_data(self, app):
+        test_artifact.refresh_from_db()
+        assert test_artifact.finalized
+
+    def test_patch_finalized__cannot_revert(self, app):
+        test_artifact, test_auth, _ = configure_test_preconditions()
+        test_artifact.finalized = True
+        test_artifact.save()
+
+        payload = make_patch_payload(test_artifact, is_finalized=False)
+        resp = app.patch_json_api(make_api_url(test_artifact), payload, auth=test_auth, expect_errors=True)
+
+        assert resp.status_code == 409
+        error_info = resp.json['errors'][0]
+        assert error_info['source']['pointer'] == '/data/attributes/finalized'
+
+    @pytest.mark.parametrize(
+        'missing_pid, missing_resource_type, expected_error_source',
+        [
+            (True, False, '/data/attributes/pid'),
+            (False, True, '/data/attributes/resource_type'),
+            (True, True, '/data/attributes/'),
+        ]
+    )
+    def test_patch_finalized__missing_required_fields(
+        self, app, missing_pid, missing_resource_type, expected_error_source
+    ):
+        test_artifact, test_auth, _ = configure_test_preconditions()
+        if missing_pid:
+            identifier = test_artifact.identifier
+            identifier.value = ''
+            identifier.save()
+        if missing_resource_type:
+            test_artifact.artifact_type = ArtifactTypes.UNDEFINED
+        test_artifact.finalized = False
+        test_artifact.save()
+
+        payload = make_patch_payload(test_artifact, is_finalized=True)
+        resp = app.patch_json_api(make_api_url(test_artifact), payload, auth=test_auth, expect_errors=True)
+
+        assert resp.status_code == 409
+        error_info = resp.json['errors'][0]
+        assert error_info['source']['pointer'] == expected_error_source
 
+    def test_patch_finalized__patch_is_atomic(self, app):
+        test_artifact, test_auth, _ = configure_test_preconditions()
+        test_artifact.artifact_type = ArtifactTypes.UNDEFINED
+        test_artifact.save()
+
+        original_title = test_artifact.title
+        original_description = test_artifact.description
+        original_identifier = test_artifact.identifier
+
+        payload = make_patch_payload(
+            test_artifact,
+            new_title='Some new name',
+            new_description='Some new desciprtion',
+            new_pid='Some new pid',
+            new_resource_type=ArtifactTypes.UNDEFINED,
+            is_finalized=True
+        )
+
+        app.patch_json_api(make_api_url(test_artifact), payload, auth=test_auth, expect_errors=True)
+        test_artifact.refresh_from_db()
+
+        assert test_artifact.title == original_title
+        assert test_artifact.description == original_description
+        assert test_artifact.artifact_type == ArtifactTypes.UNDEFINED
+        assert test_artifact.identifier == original_identifier
+        assert test_artifact.identifier.value == original_identifier.value
+        assert not test_artifact.finalized
```

Comment:
So `finalized` has to be PATCHed by itself?

Reply:
no, this was just the clearest way to show that a failed PATCH reverts all changes